### ISSUE TITLE
single-character EGCs could be inlined in struct cell->gcluster

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#830]
Right now, we only inline ASCII characters, i.e. those encoding to a single UTF-8 byte of 0x00..0x7f. Otherwise, the gcluster is taken as an offset into the egctable, where the EGC is stored as a nul-terminated string. It works well enough.

The vast majority of non-ASCII EGCs are likely to be a single Unicode character mapping to between 2 and 4 UTF-8 bytes (the 21-bit character space already maps to 4 bytes, e.g. `U+1F90C PINCHED FINGERS f0 9f a4 8c`). In the case of a single Unicode character, we can always fit it into the 32-bit `gcluster`, evading the egcplane spillover.

But then what about multicharacter EGCs? Easy: in the case of an ASCII EGC, we're wasting three bytes. All we need is a sentinel ASCII value which we don't actually allow through to the display. 0x00 is already serving a sentinel purpose (transparent glyph channel). We can hijack 0x01 for this purpose. If the first byte of `gcluster` is 0x01, we can interpret the least significant 24 bits as an index into the egcpool. If 16MB is not large enough for an egcpool, we can have the 24-bit values index into a pointer table, whatever (we've addressed this before, see e.g. #425).

To do this safely, we'll want to exclude 0x01 from ever getting into a `cell`, which would mean guarding `CELL_SIMPLE_INITIALIZER`, `cell_load_simple()`, and `cell_load_egc()` from a direct load of 0x01. If a user does it manually, they deserve what's coming to them. As Archimedes said, "do not disturb my [cells]!"

**Comment by dankamongmen** [comment]
Are we certain we want to store the (up to) 4-byte UTF-8 encoding, as opposed to the (up to) 21-bit UTF-32? Probably, since we always deal in UTF-8....but it does mean that if Unicode expanded further, up to the theoretical 6-byte max, we'd have to do something different...

well, if that unpleasant day should ever come, we'll switch to UTF-32 then.

**Comment by dankamongmen** [comment]
This will be a big change, so let's walk through what needs to be done:

* `cell_simple_c()` should probably return true for anything inlined. so it becomes `return c->gcluster != 0x1`.
  * callers need to be checked that they're not assigning `gcluster` to anything narrower than 32 bits
  * ...and not using it with ctype-style functions or e.g. `fputc()`
  * come to think of it, how **do** we print one of these? it's utf-8, but not nul-terminated, and we don't know the length...`%.4s`? might lead to significant slowdown...
* `cell_egc_idx()` becomes `c->gcluster >> 8`
* `egcpool_extended_gcluster()` is unchanged
* `cell_duplicate_far()` returns `strnlen(g->cluster, 4)` for inlined EGCs
  * ... and sets `targ->gcluster = eoffset << 8 + 1` for non-inlined
* `pool_egc_copy()` must now `malloc()` up to 5 bytes, `memcpy()` in `gcluster`, and nul-terminate
* `cell_load_direct()` must now inline anything of 4 bytes or less, instead of 1
  * ... and must `memcpy()` in the bytes, then nul-terminate if < 4
...

**Comment by dankamongmen** [comment]
I've not got this all working (we break in `normal`, and generate crap in a few places), but initial results are **very** promising, with our first-ever post of 6K+ FPS(!):

```
              runtime│ frames│output(B)│rendering│ %r│    FPS│TheoFPS║
══╤═════════╤════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╣
 0│ whiteout│  10.00s│  54596│  10.62Mi│    8.96s│ 89│ 5459.4│ 6095.7║
══╧═════════╪════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╝
            │  10.00s│  54596│  10.62Mi│    8.96s│ 89│ 5459.4│

```

**Comment by dankamongmen** [comment]
alacritty 78x104 with (`normal` and `fallin'` fail, and have been left out):

```
9119 renders, 8.31s total (177.72µs min, 6.79ms max, 911.18µs avg)
442.38MiB total (8.00B min, 300.65KiB max, 49.7KiB avg)
1097.5 theoretical FPS, 0 failed renders
RGB emits:elides: def 463288:8951 fg 10932747:5162223 bg 12920612:3581105
 Elide rates: 1.90% 32.07% 21.70%
Cell emits:elides: 16538761/57570021 (77.68%)

              runtime│ frames│output(B)│rendering│ %r│    FPS│TheoFPS║
══╤═════════╤════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╣
 0│    intro│  8.19ms│      2│ 246.85Ki│   5.98ms│ 73│  244.3│  334.2║
 1│     xray│   1.10s│    302│  18.63Mi│ 366.85ms│ 33│  273.8│  823.2║
 2│    eagle│   2.47s│    206│  27.49Mi│ 599.27ms│ 24│   83.3│  343.7║
 3│      zoo│111.15ms│    368│ 941.89Ki│  89.17ms│ 80│ 3310.9│ 4126.8║
 4│    yield│   1.58s│    393│ 890.20Ki│  80.30ms│  5│  248.4│ 4894.2║
 5│   dragon│ 36.81ms│     19│ 407.03Ki│  10.60ms│ 28│  516.2│ 1791.8║
 6│    trans│  6.19ms│      9│ 241.48Ki│   4.21ms│ 68│ 1455.0│ 2135.4║
 7│  highcon│   1.01s│   4004│   4.72Mi│ 951.62ms│ 94│ 3960.9│ 4207.6║
 8│      box│107.44ms│     51│   6.10Mi│  92.61ms│ 86│  474.7│  550.7║
 9│   chunli│ 66.31ms│     53│   1.29Mi│  26.46ms│ 39│  799.3│ 2002.9║
10│ mojibake│309.69ms│    496│  18.79Mi│ 298.70ms│ 96│ 1601.6│ 1660.5║
11│     grid│   2.55s│    768│ 184.56Mi│    2.45s│ 96│  301.4│  313.5║
12│     reel│  6.83ms│      1│ 182.04Ki│   4.36ms│ 63│  146.4│  229.3║
13│ whiteout│ 12.81ms│     14│ 149.46Ki│   4.47ms│ 34│ 1092.9│ 3130.2║
14│ uniblock│ 27.68ms│     70│ 926.07Ki│  19.15ms│ 69│ 2529.2│ 3655.2║
15│     view│   3.92s│    753│ 129.17Mi│    2.37s│ 60│  191.9│  318.0║
16│    luigi│ 58.60ms│    137│   2.34Mi│  47.93ms│ 81│ 2337.8│ 2858.1║
17│  sliders│ 57.55ms│    201│   1.17Mi│  48.51ms│ 84│ 3492.8│ 4143.8║
18│   jungle│ 38.87ms│      2│ 120.33Ki│   6.23ms│ 16│   51.4│  320.9║
19│   qrcode│   1.12s│   1024│  17.54Mi│ 361.93ms│ 32│  917.8│ 2829.3║
20│    outro│   1.63s│    244│  26.55Mi│ 471.71ms│ 28│  149.9│  517.3║
══╧═════════╪════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╝
            │  16.23s│   9117│ 442.37Mi│    8.31s│ 51│  561.7│
```

without:

```
              runtime│ frames│output(B)│rendering│ %r│    FPS│TheoFPS║
══╤═════════╤════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╣
 0│    intro│  8.81ms│      2│ 246.85Ki│   6.26ms│ 71│  227.1│  319.5║
 1│     xray│   1.18s│    302│  18.63Mi│ 408.34ms│ 34│  255.9│  739.6║
 2│    eagle│   2.50s│    206│  27.49Mi│ 625.97ms│ 25│   82.4│  329.1║
 3│      zoo│114.17ms│    368│ 945.62Ki│  92.28ms│ 80│ 3223.3│ 3988.0║
 4│    yield│   1.59s│    396│ 889.01Ki│  92.96ms│  5│  248.4│ 4259.9║
 5│   dragon│ 38.62ms│     19│ 407.01Ki│  11.48ms│ 29│  492.0│ 1655.5║
 6│    trans│  6.63ms│      9│ 246.38Ki│   4.50ms│ 67│ 1358.3│ 2001.7║
 7│  highcon│   1.01s│   4004│   4.68Mi│ 948.03ms│ 93│ 3964.9│ 4223.5║
 8│      box│142.31ms│     51│   6.11Mi│ 120.76ms│ 84│  358.4│  422.3║
 9│   chunli│ 70.58ms│     53│   1.31Mi│  29.16ms│ 41│  751.0│ 1817.3║
10│ mojibake│433.25ms│    496│  18.07Mi│ 418.82ms│ 96│ 1144.8│ 1184.3║
11│     grid│   3.05s│    768│ 184.55Mi│    2.82s│ 92│  252.1│  272.1║
12│     reel│  6.31ms│      1│ 181.93Ki│   3.07ms│ 48│  158.4│  325.7║
13│ whiteout│ 16.33ms│     14│ 155.43Ki│   6.59ms│ 40│  857.1│ 2123.1║
14│ uniblock│ 29.68ms│     70│ 919.64Ki│  21.34ms│ 71│ 2358.4│ 3280.4║
15│     view│   4.13s│    753│ 129.16Mi│    2.50s│ 60│  182.4│  301.5║
16│    luigi│ 69.10ms│    137│   2.34Mi│  58.24ms│ 84│ 1982.6│ 2352.3║
17│  sliders│ 67.13ms│    201│   1.16Mi│  57.31ms│ 85│ 2994.0│ 3507.5║
18│   jungle│ 38.63ms│      2│ 122.53Ki│   7.14ms│ 18│   51.8│  279.9║
19│   qrcode│   1.18s│   1024│  18.22Mi│ 404.91ms│ 34│  869.5│ 2528.9║
20│    outro│   1.74s│    244│  26.54Mi│ 562.08ms│ 32│  139.9│  434.1║
══╧═════════╪════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╝
            │  17.42s│   9120│ 442.27Mi│    9.20s│ 52│  523.5│
```

these are big gains across the board =] =] =] only `highcon` slowed down at all, and that barely. some are seeing 33%--40% jumps. this is fucking huge.

**Comment by dankamongmen** [comment]
With the successful resolution of #884, we open up 8 bits in the attrword for use as a last-ditch NUL terminator. This ought allow us  to recoup the speed lost in extraction, and still evidence the 10%+ perf gain seen elsewhere (more like 40% in some places). I'm going to try and get this tested and in today. 1.7.0 is shaping up to be the best release ever.

**Comment by dankamongmen** [comment]
We're now running everything properly with inline egcs. Need to do perf evaluation afresh, and then see what we can claw back based on the above, but I think we have a dance-off winner.

**Comment by dankamongmen** [comment]
Before: 18.91, 18.80, 17.64 (less is better)
After:  19.07, 18.94, 18.89

well, that's ass. let's figure out what made us so much slower between happy halcyon days of yore and this year of our fecal lord.

**Comment by dankamongmen** [comment]
This was the culprit:

```
￼      if(ncfputc(c->gcluster, out) == EOF){
```

was replaced with 

```
￼    }else if(fprintf(out, "%.4s", (const char*)&c->gcluster) < 0){
```

i forgot Dank's First Rule of Perf: fuck formatted output. yanking this boosts us back well up over the original. and this will be safe once we lock down the first byte of attrword. we're back in business, bitches!

https://www.youtube.com/watch?v=pAgnJDJN4VA

**Comment by dankamongmen** [comment]
with these changes, we're reliably sub-18s on a `notcurses-demo -c -k -p ../data` in an alacritty 78x104 on my workstation (we were supra-19s before)

**Comment by dankamongmen** [comment]
Before: 19.34, 18.81, 19.12
After: 17.91, 17.87, 17.90

now **that's** more like it! let's watch a regular delay (`-d1`) run and see what's to be seen, but i think this is going in tonight.

**Comment by dankamongmen** [comment]
Before:

```
              runtime│ frames│output(B)│rendering│ %r│    FPS│TheoFPS║
══╤═════════╤════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╣
 0│    intro│   3.03s│    296│  11.20Mi│ 365.54ms│ 12│   97.7│  809.8║
 1│     xray│   5.04s│    302│  18.57Mi│ 510.98ms│ 10│   59.9│  591.0║
 2│    eagle│   4.58s│    311│  27.31Mi│ 662.78ms│ 14│   67.9│  469.2║
 3│      zoo│   8.14s│    330│ 917.17Ki│ 156.66ms│  1│   40.5│ 2106.4║
 4│ allglyph│   1.65s│   3053│  82.69Mi│    1.49s│ 90│ 1851.7│ 2048.7║
 5│    yield│   6.95s│    827│ 921.73Ki│ 285.04ms│  4│  118.9│ 2901.3║
 6│   dragon│   2.44s│     57│ 410.11Ki│  30.94ms│  1│   23.3│ 1842.1║
 7│    trans│   8.06s│    911│   1.57Mi│ 355.27ms│  4│  113.1│ 2564.3║
 8│  highcon│   1.11s│   4004│   4.28Mi│ 953.39ms│ 86│ 3616.8│ 4199.8║
 9│   normal│   6.45s│    312│   5.80Mi│ 257.40ms│  3│   48.3│ 1212.1║
10│      box│   3.26s│     51│   5.97Mi│ 164.30ms│  5│   15.7│  310.4║
11│   chunli│   5.45s│    106│   1.31Mi│  71.74ms│  1│   19.4│ 1477.5║
12│ mojibake│  13.75s│    992│  17.96Mi│ 888.67ms│  6│   72.1│ 1116.3║
13│     grid│   3.04s│    768│ 183.02Mi│    2.76s│ 90│  252.2│  277.9║
14│     reel│   5.00s│     47│   2.79Mi│  86.55ms│  1│    9.4│  543.1║
15│ whiteout│  10.01s│  34100│  13.94Mi│    7.40s│ 73│ 3408.1│ 4610.1║
16│ uniblock│  14.11s│    210│ 985.91Ki│ 101.76ms│  0│   14.9│ 2063.6║
17│     view│  15.43s│   1533│ 127.90Mi│    3.31s│ 21│   99.4│  463.0║
18│    luigi│   1.78s│    274│   2.35Mi│ 156.11ms│  8│  153.5│ 1755.2║
19│  sliders│   5.28s│    539│   1.60Mi│ 199.25ms│  3│  102.2│ 2705.2║
20│  fallin'│   7.83s│   5738│   3.93Mi│    2.02s│ 25│  733.0│ 2836.8║
21│   jungle│   5.04s│    302│ 267.78Ki│ 113.09ms│  2│   59.9│ 2670.3║
22│   qrcode│   1.20s│   1024│  18.21Mi│ 379.51ms│ 31│  851.6│ 2698.2║
23│    outro│  10.05s│    727│  45.01Mi│    1.36s│ 13│   72.3│  533.4║
══╧═════════╪════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╝
            │ 148.69s│  56814│ 578.84Mi│   24.09s│ 16│  382.1│
```

after:

```              runtime│ frames│output(B)│rendering│ %r│    FPS│TheoFPS║
══╤═════════╤════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╣
 0│    intro│   3.03s│    296│  11.20Mi│ 335.85ms│ 11│   97.8│  881.3║
 1│     xray│   5.05s│    302│  18.57Mi│ 489.57ms│  9│   59.9│  616.9║
 2│    eagle│   4.53s│    311│  27.31Mi│ 640.95ms│ 14│   68.6│  485.2║
 3│      zoo│   8.12s│    330│ 917.25Ki│ 139.69ms│  1│   40.6│ 2362.4║
 4│ allglyph│   1.50s│   3053│  82.69Mi│    1.35s│ 89│ 2029.3│ 2255.8║
 5│    yield│   6.91s│    827│ 920.84Ki│ 247.67ms│  3│  119.6│ 3339.1║
 6│   dragon│   2.44s│     57│ 410.12Ki│  27.79ms│  1│   23.4│ 2051.4║
 7│    trans│   8.04s│    915│   1.58Mi│ 334.90ms│  4│  113.8│ 2732.1║
 8│  highcon│   1.09s│   4004│   4.20Mi│ 944.35ms│ 86│ 3665.6│ 4240.0║
 9│   normal│   6.40s│    312│   5.39Mi│ 232.23ms│  3│   48.7│ 1343.5║
10│      box│   3.24s│     51│   5.96Mi│ 157.88ms│  4│   15.7│  323.0║
11│   chunli│   5.44s│    106│   1.31Mi│  63.67ms│  1│   19.5│ 1664.9║
12│ mojibake│  13.67s│    992│  17.96Mi│ 806.22ms│  5│   72.6│ 1230.4║
13│     grid│   2.52s│    768│ 182.98Mi│    2.38s│ 94│  304.2│  322.9║
14│     reel│   5.00s│     47│   2.79Mi│  80.75ms│  1│    9.4│  582.0║
15│ whiteout│  10.00s│  36844│  14.92Mi│    7.36s│ 73│ 3682.6│ 5005.9║
16│ uniblock│  14.10s│    210│ 983.77Ki│  90.99ms│  0│   14.9│ 2307.9║
17│     view│  15.42s│   1533│ 127.92Mi│    3.12s│ 20│   99.4│  491.9║
18│    luigi│   1.76s│    274│   2.35Mi│ 136.35ms│  7│  155.3│ 2009.6║
19│  sliders│   5.26s│    543│   1.61Mi│ 181.28ms│  3│  103.3│ 2995.3║
20│  fallin'│   7.80s│   5967│   3.96Mi│    1.88s│ 24│  764.6│ 3179.5║
21│   jungle│   5.04s│    302│ 267.67Ki│ 105.71ms│  2│   59.9│ 2857.0║
22│   qrcode│   1.17s│   1024│  18.03Mi│ 366.74ms│ 31│  877.9│ 2792.2║
23│    outro│  10.04s│    727│  45.02Mi│    1.34s│ 13│   72.4│  543.3║
══╧═════════╪════════╪═══════╪═════════╪═════════╪═══╪═══════╪═══════╝
            │ 147.60s│  59795│ 579.16Mi│   22.81s│ 15│  405.1│
```

**Comment by dankamongmen** [comment]
Looks like about a ~6% improvement in FPS, which lines up pretty closely with our ~5% cut in runtime with `-d`.

Well, in the words of compiler authors everywhere, "2% perf gains are why they pay me the big bucks."

Let's rigourize this fucker and toss it in. Tally-ho!

https://www.youtube.com/watch?v=lVehcuJXe6I

**Comment by dankamongmen** [comment]
On my laptop using `xfce4-terminal` + i915, runtime with `-d0` went from 67.54 to 64.78 (155->163 FPS), just about the same 5% gain.

**Comment by dankamongmen** [comment]
I think we got rid of all the `strlen()`s which had been added, but if we needed such a thing, we could encode the following:

1: egc spill
2: two-byte UTF8
3: three-byte UTF8
< 128: single-byte ASCII
>= 128: four-byte UTF8

and that ought make short work of it.

**Comment by dankamongmen** [comment]
As could be predicted, the biggest speedup is on `grid`, where we (a) use entirely single-Unicode, supra-ASCII EGCs, (b) fill the screen with them, and (c) do very little computation, just pump out the I/O. That gets about a 20% speedup. `allglyphs` is in the same boat (though libfreetype contributes a good deal of delay), and picks up a 10% speedup. Nice, as these were some of our most intense demos.

The timings on `view` and `xray` are almost exactly the same (-.01s, +.01s). This surprises me, since that's video playback, and thus blitting, and thus everything ought be single-Unicode supra-ASCII (henceforth SUSA) EGCs. I suppose actual video decoding is dominating that time, especially at `-d0`.

It doesn't look like anything slowed down beyond noise.

Great. This both is a nice win in and of itself, and frankly seems evidence that our core overall is pretty lean. I doubt further wins exist in the encoding / lookup layers. Our next perf step must be #774 (if we need a perf step. We're by any measure fast as fuck).

**Comment by dankamongmen** [comment]
Added the desired unit tests (see #887), fixed up a bug that would have bit us on big-endian systems (I really, really, really would like to eliminate the endian awareness, ugh), and I think this is ready.

**Comment by dankamongmen** [comment]
Hrmmm. So long as we threw a `((packed))` on the `cell` structure, what if we broke the `uint32_t attrword` out into

```
uint8_t backstop; // must always be zero, keep your grubby mitts off me
uint8_t reserved; // should be zero
uint16_t attrword; // actual attrword
```

i mean....that ought work just fine, right? and eliminate our endianness issues? i think it might! hallelujah! olé, olé, olé!

@joseluis do you see such a plan possibly causing problems for our Rust? i imagine that, if anything, it might make things easier (less within-identifier bit twiddling)?

i'm gonna see what breaks if i do this, and if it'll fly, this makes me feel 100% better about this change, which will really become pure unadulterated win.

**Comment by dankamongmen** [comment]
> Hrmmm. So long as we threw a `((packed))` on the `cell` structure, what if we broke the `uint32_t attrword` out into
> 
> ```
> uint8_t backstop; // must always be zero, keep your grubby mitts off me
> uint8_t reserved; // should be zero
> uint16_t attrword; // actual attrword
> ```
> 
> i mean....that ought work just fine, right? and eliminate our endianness issues? i think it might! hallelujah! olé, olé, olé!
> 
> @joseluis do you see such a plan possibly causing problems for our Rust? i imagine that, if anything, it might make things easier (less within-identifier bit twiddling)?
> 
> i'm gonna see what breaks if i do this, and if it'll fly, this makes me feel 100% better about this change, which will really become pure unadulterated win.

(for those playing along at home, the only reason the attrword was 32 bits instead of 16 was to make the `cell` an even 128 bits, and to ensure good alignment for `channels`).

**Comment by dankamongmen** [comment]
Hrmmm do we even really want to `((pack))`? If the compiler thinks it better to align a `uint16_t` to a 4-byte boundary (and thus throw the whole struct stride alignment out of whack), ought we not trust it? After all, it'll need to pad the `struct` itself up to at least a 4-byte boundary (total of 160 bits, yuck) to be correct, and presumably it knows that cost and has made this decision purposefully?

Let's try it without `((packed))` and see what we're looking at on a few platforms.

**Comment by dankamongmen** [comment]
If we go this route, we probably ought rename `attrword` to `styles` or `styling`.

**Comment by dankamongmen** [comment]
OK, doing this without compiler warnings is going to change a lot of function signatures. I think I'm going to merge #887 now, mark this fixed, and then create a new issue + PR for eliminating the endianness crap.

**Comment by dankamongmen** [comment]
> To do this safely, we'll want to exclude 0x01 from ever getting into a `cell`, which would mean guarding `CELL_SIMPLE_INITIALIZER`, `cell_load_simple()`, and `cell_load_egc()` from a direct load of 0x01. If a user does it manually, they deserve what's coming to them. As Archimedes said, "do not disturb my [cells]!"


To close the loop on this, it ought be pointed out that Archimedes (supposedly) said this right before having a spear plunged between his shoulder blades.

> The last words attributed to Archimedes are "Do not disturb my circles", a reference to the circles in the mathematical drawing that he was supposedly studying when disturbed by the Roman soldier. This quote is often given in Latin as "Noli turbare circulos meos," but there is no reliable evidence that Archimedes uttered these words and they do not appear in the account given by Plutarch. Valerius Maximus, writing in Memorable Doings and Sayings in the 1st century AD, gives the phrase as "…sed protecto manibus puluere 'noli' inquit, 'obsecro, istum disturbare'" ("…but protecting the dust with his hands, said 'I beg of you, do not disturb this'"). The phrase is also given in Katharevousa Greek as "μὴ μου τοὺς κύκλους τάραττε!" (Mē mou tous kuklous taratte!)

**Comment by joseluis** [comment]
> @joseluis do you see such a plan possibly causing problems for our Rust? i imagine that, if anything, it might make things easier (less within-identifier bit twiddling)?

I don't think so, let's do it
